proposals/call-tags/Overview.md

```diff
@@ -4,183 +4,254 @@
+For example, if one specifies no canonical call tags and only non-exported call tags, then one can be guaranteed that the function is only indirectly called by this module instance, reducing the possibility of unintended backdoors.
 
-One way to implement `call_indirect` is to push all the arguments onto the stack and then to push a signature descriptor onto the stack; the callee then pops that descriptor and compares it for (bitwise) equality with the callee's own signature descriptor.
-If the two match, then the invariants guaranteed by the static type-checker ensure the stack has the arguments expected by the function (and similarly that the caller is expecting the values that will be returned by the function), so the function call proceeds.
-If they don't match, then it traps.
-(This is how SpiderMonkey currently implements `call_indirect`.)
+Notice that a `funcref` can handle multiple call tags, even multiple canonical call tags.
+This is because dynamic typing is implemented by the call*ee* rather than the call*er*, which experiments suggest can be implemented much more efficiently *in the case of function calls*.
+This also opens up the opportunity for a `funcref` to actually provide different behavior for different call tags.
+So this proposal includes a `func_switch ($call_tag $func)*` construct in the code section that creates a new `funcref` that tail-calls the `$func` corresponding to the used call tag.
+[Efficient implementation of Java interfaces: Invokeinterface considered harmless](https://dl.acm.org/doi/10.1145/504282.504291) used this to implement Java interface methods and found the technique to make interface-method dispatch typically as fast as class-method dispatch, and our recent experiments confirmed that this behavior still holds in modern hardware.
 
-We can solve all the problems above by generalizing this process *call tags*.
-That is, a call tag is a generalization of these signature descriptors.
+Given all that, what happens when the specified call tag is *not* handled by the `funcref` at hand?
+For canonical call tags, the answer is simply that the program traps.
+But when creating call tags with `call_tag.new $functype` (in the Tags section), one can also specify a `$func` to use as its "fall-back" handler.
+This `$func` must have the same signature as `$functype` *except* also accepting an additional `funcref` so that we can pass the fall-back handler the specific `funcref` that did *not* recognize the call tag.
+If engines are careful with their calling conventions, the fall-back can be switched to with just a jump instruction, making it a very efficient backup.
+[Transitioning from Structural to Nominal Code with Efficient Gradual Typing](https://www.cs.cornell.edu/~ross/publications/monnom/monnom-oopsla21.pdf) used this to implement a dynamically typed language and found it to work quite well.
```

Comment:
This link gets used a lot. Can you perhaps give it (and the paper) a short name?